evm/src/lib.rs

```diff
@@ -1,3 +1,153 @@
+//! aggregation levels (see below). If one were to miss the public values, they are also retrievable directly
+//! from the proof's encoded public inputs, as such:
+//!
+//! ```ignore
+//! let public_values = PublicValues::from_public_inputs(&proof.public_inputs);
+//! ```
+//!
+//! ## Aggregation proofs
+//!
+//! Because the plonky2 zkEVM generates proofs on a transaction basis, we then need to aggregate them for succinct
+//! verification. This is done in a binary tree fashion, where each inner node proof verifies two children proofs,
+//! through the [prove_aggreg](AllRecursiveCircuits::prove_aggreg) method.
+//! Note that the tree does *not* need to be complete, as this aggregation process can take as inputs both regular
+//! transaction proofs and aggregation proofs. We only need to specify for each child if it is an aggregation proof
+//! or a regular one.
+//!
+//! ```ignore
+//! let (proof_1, pv_1) =
+//!     prover_state.prove_root(all_stark, config, inputs_1, &mut timing, None);
+//! let (proof_2, pv_2) =
+//!     prover_state.prove_root(all_stark, config, inputs_2, &mut timing, None);
+//! let (proof_3, pv_3) =
+//!     prover_state.prove_root(all_stark, config, inputs_2, &mut timing, None);
+//!
+//! // Now aggregate proofs for txn 1 and 2.
+//! let (agg_proof_1_2, pv_1_2) =
+//!     prover_state.prove_aggreg(false, proof_1, pv_1, false, proof_2, pv_2);
+//!
+//! // Now aggregate the newly generated aggregation proof with the last regular txn proof.
+//! let (agg_proof_1_3, pv_1_3) =
+//!     prover_state.prove_aggreg(true, agg_proof_1_2, pv_1_2, false, proof_3, pv_3);
+//! ```
+//!
+//! **Note**: The proofs provided to the [prove_aggreg](AllRecursiveCircuits::prove_aggreg) method *MUST* have contiguous states.
+//! Trying to combine `proof_1` and `proof_3` from the example above would fail.
+//!
+//! ## Block proofs
+//!
+//! Once all transactions of a block have been proven and we are left with a single aggregation proof and its public values,
+//! we can then wrap it into a final block proof, attesting validity of the entire block.
+//! This [prove_block](AllRecursiveCircuits::prove_aggreg) method accepts an optional previous block proof as argument,
+//! which will then try combining the previously proven block with the current one, generating a validity proof for both.
+//! Applying this process from genesis would yield a single proof attesting correctness of the entire chain.
+//!
+//! ```ignore
+//! let previous_block_proof = { ... };
+//! let (block_proof, block_public_values) =
+//!     prover_state.prove_block(Some(&previous_block_proof), &agg_proof, agg_pv)?;
+//! ```
+//!
+//! ### Checkpoint heights
+//!
+//! Because the process of always providing a previous block proof when generating a proof for the current block may yield some
+//! undesirable issues, the plonky2 zk-EVM supports checkpoint heights. At given block heights, the prover does not have to pass
```

Comment:
> undesirable issues

Such as?

Reply:
If proving is decentralized for instance, and a prover falls behind. Or if a prover was to crash and lose its DB, with previously generated proofs, it would need to start again from genesis.

Reply:
decentralized? You mean distributed yeah?

So by "undesirable issues" you mean "avoiding restarting from scratch after a crash or missing previous data"? Might as well write that then! :)

Do you think we should mention that using the checkpoint feature is probably not a good idea for production systems (but awesome for development)?

Reply:
I did mean decentralized. We could picture a rewarding scenario where first prover to submit a valid proof from checkpoint A to checkpoint B is rewarded.

Even in the case of centralized proving, the checkpoint feature is useful for the other issues I mentioned.

Reply:
Checkpointing can really be seen as skipping unnecessary work. Once you've submitted your proof on your settlement layer, and it's been verified and approved, your state up to block N is verified. Hence no need to keep track to what previously happen, and locally do a proof from block N to block N+c for some constant c. Proper connection between the two proofs should be done on the settlement layer, to ensure that the starting state of the incoming proof was the final state of the previously submitted proof (much like what's being done when aggregating transactions, except that this is done internally in the circuits).